### PR TITLE
[Alerting]: Update scheduler to evaluate rules created by the unified API

### PR DESCRIPTION
**Pull request by papagian**
Rules are fetched, evaluated and results are handled by the state tracker

```
DBUG[04-01|12:31:36] alert rules fetched                      logger=ngalert count=2
DBUG[04-01|12:31:36] alert rule routine started               logger=ngalert key="{orgID: 1, UID: F01iqH_Mzz}"
DBUG[04-01|12:31:36] alert rule routine started               logger=ngalert key="{orgID: 1, UID: K0JmqNlGz}"
DBUG[04-01|12:31:36] new alert rule version fetched           logger=ngalert title="prom query with SSE" key="{orgID: 1, UID: K0JmqNlGz}" version=1
{"@level":"debug","@message":"expression datasource query (seriesSet)","@timestamp":"2021-04-01T12:31:36.211955+03:00","query":"query"}
INFO[04-01|12:31:36] alert rule result                        logger=ngalert title="prom query with SSE" key="{orgID: 1, UID: K0JmqNlGz}" attempt=0 now=2021-04-01T12:31:36+0300 duration=5.000781ms instance= state=Normal
INFO[04-01|12:31:36] state tracker processing evaluation results logger=ngalert uid=K0JmqNlGz resultCount=1
DBUG[04-01|12:31:36] adding new alert state cache entry       logger=ngalert cacheId="K0JmqNlGz " state=Normal evaluatedAt="2021-04-01 12:31:36.205216 +0300 EEST m=+12.812646671"
DBUG[04-01|12:31:36] setting alert state                      logger=ngalert uid=K0JmqNlGz
DBUG[04-01|12:31:36] no state transition                      logger=ngalert cacheId="K0JmqNlGz " state=Normal
DBUG[04-01|12:31:36] returning changed states to scheduler    logger=ngalert count=0
DBUG[04-01|12:31:37] Scheduling update                        logger=alerting.scheduler ruleCount=5
DBUG[04-01|12:31:41] new alert rule version fetched           logger=ngalert title="reduced testdata query" key="{orgID: 1, UID: F01iqH_Mzz}" version=1
{"@level":"debug","@message":"expression datasource query (seriesSet)","@timestamp":"2021-04-01T12:31:41.208267+03:00","query":"A"}
INFO[04-01|12:31:41] alert rule result                        logger=ngalert title="reduced testdata query" key="{orgID: 1, UID: F01iqH_Mzz}" attempt=0 now=2021-04-01T12:31:36+0300 duration=1.219014ms instance= state=Alerting
INFO[04-01|12:31:41] state tracker processing evaluation results logger=ngalert uid=F01iqH_Mzz resultCount=1
DBUG[04-01|12:31:41] adding new alert state cache entry       logger=ngalert cacheId="F01iqH_Mzz " state=Alerting evaluatedAt="2021-04-01 12:31:36.205216 +0300 EEST m=+12.812646671"
DBUG[04-01|12:31:41] setting alert state                      logger=ngalert uid=F01iqH_Mzz
DBUG[04-01|12:31:41] no state transition                      logger=ngalert cacheId="F01iqH_Mzz " state=Alerting
DBUG[04-01|12:31:41] returning changed states to scheduler    logger=ngalert count=0
```

### CODE REVIEW

**Review by gotjosh — APPROVED** [review]

File: pkg/services/ngalert/schedule/fetcher.go [hunk 1]
```diff
@@ -6,9 +6,9 @@ import (
...
 )
 
-func (sch *schedule) fetchAllDetails(now time.Time) []*models.AlertDefinition {
-	q := models.ListAlertDefinitionsQuery{}
-	err := sch.store.GetAlertDefinitions(&q)
+func (sch *schedule) fetchAllDetails(now time.Time) []*models.AlertRule {
+	q := models.ListAlertRulesQuery{}
+	err := sch.ruleStore.GetAlertRulesForScheduling(&q)
 	if err != nil {
 		sch.log.Error("failed to fetch alert definitions", "now", now, "err", err)
```
[the remark this answers is not in the document]
Comment (papagian) [reply pkg/services/ngalert/schedule/fetcher.go:13]: `now` is actually the scheduler tick that can differ from the logging time (and I had it there for debugging).
https://github.com/grafana/grafana/blob/01c46fcac024b9b8d0d08f3e1ee36e8f43c11ca8/pkg/services/ngalert/schedule/schedule.go#L247
But I agree that there is not point including it in the error log.

**Comment by papagian** [comment]
Now that we have introduce [AlertRule.ExecErrState](https://github.com/grafana/grafana/blob/b714cdf395d1c89eaff81b2eb4ee39298bb8dfec/pkg/services/ngalert/models/alert_rule.go#L54) I have to change this and save the error state instead of returning
https://github.com/grafana/grafana/blob/01c46fcac024b9b8d0d08f3e1ee36e8f43c11ca8/pkg/services/ngalert/schedule/schedule.go#L73-L79

**Comment by kylebrandt** [comment]
Running things :-) 

I am current getting:

```
=ngalert title="prom query with SSE classic condition" key="{orgID: 1, UID: X73TOSlMz}" attempt=0 now=2021-04-01T07:55:56-0400 duration=6.019991ms instance= state=NoData
EROR[04-01|07:55:56] failed saving alert instance             logger=ngalert title="prom query with SSE classic condition" key="{orgID: 1, UID: X73TOSlMz}" attempt=0 now=2021-04-01T07:55:56-0400 instance= state=NoData error="alert instance is invalid because the state 'NoData' is invalid"
```

The Data in alert rule is:

```
[
  {
    "refId": "A",
    "queryType": "",
    "relativeTimeRange": {
      "from": 300,
      "to": 0
    },
    "model": {
      "datasource": "gdev-prometheus",
      "datasourceUid": "000000002",
      "expr": "avg (node_network_transmit_bytes_total) by (instance,device)",
      "hide": false,
      "interval": "",
      "intervalMs": 1000,
      "legendFormat": "",
      "maxDataPoints": 100,
      "refId": "A"
    }
  },
  {
    "refId": "B",
    "queryType": "",
    "relativeTimeRange": {
      "from": 0,
      "to": 0
    },
    "model": {
      "conditions": [
        {
          "evaluator": {
            "params": [
              0
            ],
            "type": "gt"
          },
          "operator": {
            "type": "and"
          },
          "query": {
            "params": [
              "A"
            ]
          },
          "reducer": {
            "params": [],
            "type": "avg"
          },
          "type": "query"
        }
      ],
      "datasource": "__expr__",
      "datasourceUid": "-100",
      "hide": false,
      "intervalMs": 1000,
      "maxDataPoints": 100,
      "refId": "B",
      "type": "classic_conditions"
    }
  }
]
```

**Comment by papagian** [comment]
> Running things :-)
> 
> I am current getting `state=NoData error="alert instance is invalid because the state 'NoData' is invalid"`:

Good catch; that's happens because [InstanceStateType](https://github.com/grafana/grafana/blob/4ce0a49eac03e8f37914a84c8bfaeac573582e93/pkg/services/ngalert/models/instance.go) is missing some states. It needs to sync with eval.State
https://github.com/grafana/grafana/blob/b1cb74c0c9450449ad13dc0405cfca87ad55b6e7/pkg/services/ngalert/eval/eval.go#L68-L86
I'm fixing it

**Review by kylebrandt — APPROVED** [review]

**Review by gotjosh — CHANGES_REQUESTED** [review]
I'm going to briefly block this work as there's more we need to do here I'll post the details shortly.

**Review by gotjosh — APPROVED** [review]

**Pull request merged**